### ISSUE TITLE
Contrastive loss training on multiple GPUs

### ISSUE THREAD

**Issue by zlenyk** [horovod/horovod#1797]
**Environment:**
1. Framework: (TensorFlow, Keras, PyTorch, MXNet): Pytorch
2. Framework version: 
3. Horovod version: 0.19.0
4. MPI version:
5. CUDA version: 10
6. NCCL version:
7. Python version: 3.7
8. OS and version: Ubuntu
9. GCC version:

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

Could you help me understand how one trains any type of contrastive loss in data-parallel mode on multiple GPUs? I'm using horovod for handling multi-GPUs. When doing classification training, matter is rather simple - each model output is compared to it's label. But when outputs and loss function depends on the whole batch, I'm not sure how that works.

If, for example, I'm training on 8 GPUs with BS=64, "effective batch size" will be 64*8 in classification case. But when I'm using any loss that compares outputs to other outputs in the batch, what's the point of using multiple GPUs?

Thanks!

**Comment by abditag2** [comment]
Seems that after your forward pass, you need to gather all the outputs, and labels then calculate the loss. At this point, only one of the GPUs needs to calculate the loss and disperse the results to the rest.  From the description of your problem, it does not seem that you cannot use pure data-parallel for training because you need all the batch values for every step of the training of each replica of the model. 

You might be able to use all gather and all reduce functions to perform some of the above steps and if all your GPUs are on the same machine or you have a really good network, you might see faster multi GPU training.
